src/orion/testing/algo.py

```diff
@@ -143,19 +95,69 @@ class BaseAlgoTests:
 
-    algo_name = None
-    config = {}
-    max_trials = 200
-    space = {"x": "uniform(0, 1)", "y": "uniform(0, 1)"}
+    algo_name: ClassVar[str | None] = None
+
+    algo_type: type[AlgoType]
+    """ The type of algorithm under test."""
+
+    config: ClassVar[dict] = {}
+    max_trials: ClassVar[int] = 200
+    space: ClassVar[dict] = {"x": "uniform(0, 1)", "y": "uniform(0, 1)"}
 
-    phases: ClassVar[list[TestPhase]]
+    phases: ClassVar[list[TestPhase]] = []
+    """ Test phases for the algorithms. Leave empty if the algorithm only has one phase."""
+
+    _current_phase: ClassVar[TestPhase | None] = None
 
     def __init_subclass__(cls) -> None:
-        if hasattr(cls, "phases"):
-            cls.set_phases(cls.phases)
+        # NOTE: It makes sense to run all the algo tests for each phase, since every test
+        # is expected to use `self.create_algo`.
+        if cls.phases:
+            # The first test phase should always have 0 as its n_trials, since algorithms are
+            # supposed to work starting from 0 trials.
+            assert cls.phases[0].n_trials == 0
+
+            @pytest.fixture(
+                name="phase",
+                autouse=True,
+                scope="module",
+                params=cls.phases,
+                ids=[phase.name for phase in cls.phases],
+            )
+            def phase(request, mocker):
+                test_phase: TestPhase = request.param
+                # Temporarily change the class attribute holding the current phase.
+                original_phase = cls._current_phase
+                cls._current_phase = test_phase
+                # NOTE: If we want to actually use this spy stuff, We could create a spy for each
+                # phase, and then in create_algo, after the force_observe, for each (phase, spy)
+                # pair, check that the call_count is equal to phase.n_trials - prev_phase.n_trials
+                # or something similar.
+                spy = mocker.spy(cls, test_phase.method_to_spy)
+                yield test_phase
+                cls._current_phase = original_phase
+
+            # Store it somewhere on the class so it gets included in the test scope.
+            cls.phase: pytest.fixture = staticmethod(phase)  # type: ignore
+
+        # Set the `algo_type` attribute.
+        if not hasattr(cls, "algo_type") or not cls.algo_type:
+            if not cls.algo_name:
+                raise RuntimeError(
+                    "Subclasses of BaseAlgoTests must set the algo_type or algo_name attributes."
+                )
+            cls.algo_type = orion.algo.base.algo_factory.get_class(cls.algo_name)
+        if not cls.algo_name:
+            cls.algo_name = cls.algo_type.__name__.lower()
 
+    @pytest.fixture(name="first_phase")
     @classmethod
-    def set_phases(cls, phases):
+    def first_phase(cls, _phase: TestPhase):
+        if _phase != cls.phases[0]:
+            pytest.skip(reason="Test runs only on first phase.")
```

Comment:
That will pollute the test list with a lot of skips. :thinking: It would nice if the phases could be used when a method has the argument `phase` or something like that, like the usual way of using pytest fixtures.

Reply:
Hmm. Not sure I agree.
I think all tests of a BaseMethodTests subclass should be run in all phases by default.
I see marking a test as `first_phase_only` a bit like adding an `xfail` on all the other phases. It's not something you should do often.

Reply:
All tests that are for argument values, configuration, convergence and is_done are not meant to be parametrized with phases. Depending on the algorithm that can be from about 1/5 to 1/3 of the tests.

Reply:
For convergence and is_done it should only be last phase IMO. Arguments and configuration should be first phase only.

Skipping tests is fine in general but skipping systematically 1/3 of the tests is a lot. It's not a big issue, but it will pollute the pytest report. If you can find an easy fix that would be great but otherwise we'll just leave it like that.

Reply:
Added a first_phase_only and last_phase_only fixture (and make the mentioned tests use it) in 1ce04a9a8b14b2193000c9222880632cacf4e7b2